### PR TITLE
fix #11967: flow typing nullability in pattern matches

### LINKED ISSUE #11967: Null checks with pattern matching are not reflected to the return type with `-Yexplicit-nulls`
## Compiler version

Scala 3.0.0-RC2 with `-Yexplicit-nulls`

## Minimized code

```Scala
// This works correctly
def f1(s: String | Null): String = if(s == null) "other" else s

// This fails to compile
def f(s: String | Null): String = s match {
  case null => "other"
  case s => s
}
```

## Output

```scala
  case s => s
            ^
            Found:    (s : String | Null)
            Required: String
```

## Expectation

My expectation is that what happens with `def f(s: String | Null): String = if(s == null) "other" else s` should apply to pattern matching too.

I couldn't find any open issue treating this, so I opened one.

Thank you.

### PR DESCRIPTION
**Pull request by olhotak**
Fixes #11967

### CODE REVIEW

File: tests/explicit-nulls/pos/flow-match.scala [hunk 1]
```diff
@@ -12,4 +12,36 @@ object MatchTest {
...
+    case _ => "other"
+    case s2 => s2
+    case s3 => s3
+  }
+
+  def f5(s: String | Null): String = s match {
+    case x => "other"
+    case s2 => s2
+    case s3 => s3
+  }
```
Comment (noti0na1) [inline tests/explicit-nulls/pos/flow-match.scala:46]: Can we add a test where the matching null is in the middle of the cases?

**Review by noti0na1 — APPROVED** [review]
LGTM

**Review by odersky — APPROVED** [review]
